subsys/net/ip/ipv6.c

```diff
@@ -482,17 +473,36 @@ enum net_verdict net_ipv6_input(struct net_pkt *pkt, bool is_loopback)
 		}
 	}
 
-	/* If we receive a packet with ll source address fe80: and destination
-	 * address is one of ours, and if the packet would cross interface
-	 * boundary, then drop the packet. RFC 4291 ch 2.5.6
-	 */
-	if (IS_ENABLED(CONFIG_NET_ROUTING) &&
-	    net_ipv6_is_ll_addr(&hdr->src) &&
-	    !net_ipv6_is_addr_mcast(&hdr->dst) &&
-	    !net_if_ipv6_addr_lookup_by_iface(net_pkt_iface(pkt),
-					      &hdr->dst)) {
-		ipv6_no_route_info(pkt, &hdr->src, &hdr->dst);
-		goto drop;
+	if (net_ipv6_is_addr_mcast(&hdr->dst)) {
+		/* Does the destination multicast address match one of the
+		 * registered multicast groups on the originating interface
+		 * of the packet?
+		 */
+		if (!net_if_ipv6_maddr_lookup(&hdr->dst, &pkt_iface)) {
```

Comment:
I think this is not enough. The multicast address needs to be added to the interface, but it might not yet joined to the group. We need to add here a call to `net_if_ipv6_maddr_is_joined()` and check the return value.